--- /usr/bin/dstat	2022-01-14 09:38:05.110564873 +0300
+++ /usr/bin/dstat	2022-01-14 08:43:46.702179267 +0300
@@ -547 +547 @@
-            if isinstance(self.val[name], types.ListType) or isinstance(self.val[name], types.TupleType):
+            if isinstance(self.val[name], list) or isinstance(self.val[name], tuple):
@@ -552 +552 @@
-            elif isinstance(self.val[name], types.StringType):
+            elif isinstance(self.val[name], str):
