Index: nginx/vanilla-bin/tests/syslog.t
===================================================================
--- nginx/vanilla-bin/tests/syslog.t	(revision 3565302)
+++ nginx/vanilla-bin/tests/syslog.t	(revision 3565303)
@@ -172,7 +172,7 @@
 
 
 like(get_syslog('/e'),
-	qr/nginx: \d{4}\/\d{2}\/\d{2} \d{2}:\d{2}:\d{2} \[error\]/,
+	qr/nginx: \d{4}\/\d{2}\/\d{2} \d{2}:\d{2}:\d{2}\.\d{3} \[error\]/,
 	'error_log format');
 like(get_syslog('/a_logf'), qr/nginx: \/a_logf:404$/, 'access_log log_format');
 
Only in b/nginx/vanilla-bin/tests/: test-results
