commit 5a47dc400e8b28b6b6c2a670b376aeec9c627e7e (HEAD -> STRM-4101)
author: nyoroon
date: 2021-05-25T16:53:47+03:00

    Fix tests

Index: nginx/vanilla-bin/tests/h2_server_tokens.t
===================================================================
--- nginx/vanilla-bin/tests/h2_server_tokens.t	(rev 0)
+++ nginx/vanilla-bin/tests/h2_server_tokens.t	(working copy)
@@ -106,7 +106,7 @@ like(header_server('/on/200'), qr/^$re$/, 'http2 tokens on 200');
 like(header_server('/on/404'), qr/^$re$/, 'http2 tokens on 404');
 like(body('/on/404'), $re, 'http2 tokens on 404 body');
 
-$re = qr/$re \Q($1)\E/ if $t->{_configure_args} =~ /--build=(\S+)/;
+$re = qr/$re-\d+\.yandex\.\d+/;
 
 like(header_server('/b/200'), qr/^$re$/, 'http2 tokens build 200');
 like(header_server('/b/404'), qr/^$re$/, 'http2 tokens build 404');
Index: nginx/vanilla-bin/tests/server_tokens.t
===================================================================
--- nginx/vanilla-bin/tests/server_tokens.t	(rev 0)
+++ nginx/vanilla-bin/tests/server_tokens.t	(working copy)
@@ -105,7 +105,7 @@ like(http_get_server('/on/200'), $re, 'tokens on 200');
 like(http_get_server('/on/404'), $re, 'tokens on 404');
 like(http_body('/on/404'), $re, 'tokens on 404 body');
 
-$re = qr/$re \Q($1)\E/ if $t->{_configure_args} =~ /--build=(\S+)/;
+$re = qr/$re-\d+\.yandex\.\d+/;
 
 like(http_get_server('/b/200'), $re, 'tokens build 200');
 like(http_get_server('/b/404'), $re, 'tokens build 404');
