--- cvs-1.11.23/configure.old	2022-05-26 12:28:46.181599571 +0300
+++ cvs-1.11.23/configure	2022-05-26 12:29:24.825367568 +0300
@@ -575,7 +575,7 @@
 PACKAGE_NAME='Concurrent Versions System (CVS)'
 PACKAGE_TARNAME='cvs'
 PACKAGE_VERSION='1.11.23'
-PACKAGE_STRING='Concurrent Versions System (CVS) 1.11.23'
+PACKAGE_STRING='Concurrent Versions System (CVS) 1.11.23 (patch-3)'
 PACKAGE_BUGREPORT='bug-cvs@nongnu.org'
 
 ac_unique_file="src/cvs.h"
