Index: cvs-1.11.23/src/logmsg.c
===================================================================
RCS file: /data/CVSROOT/cvs-1.11.23/src/logmsg.c,v
retrieving revision 1.1
retrieving revision 1.7
diff -u -r1.1 -r1.7
--- cvs-1.11.23/src/logmsg.c	20 May 2022 04:58:33 -0000	1.1
+++ cvs-1.11.23/src/logmsg.c	23 May 2022 18:06:35 -0000	1.7
@@ -158,21 +158,11 @@
 	    tag = xstrdup (li->tag);

 	    /* Force a new line.  */
-	    col = 70;
+        (void) fprintf (fp, "\n");
 	}

-	if (col == 0)
-	{
-	    (void) fprintf (fp, "%s\t", prefix);
-	    col = 8;
-	}
-	else if (col > 8 && (col + (int) strlen (p->key)) > 70)
-	{
-	    (void) fprintf (fp, "\n%s\t", prefix);
-	    col = 8;
-	}
-	(void) fprintf (fp, "%s ", p->key);
-	col += strlen (p->key) + 1;
+	(void) fprintf (fp, "%s\t%s\n", prefix, p->key);
+	col = 8 + strlen (p->key) + 1;
     }
     return (0);
 }
@@ -703,6 +693,12 @@
 		    (void) strcat (str_list, (li->rev_new
 					      ? li->rev_new : "NONE"));
 		    break;
+        case 't':
+            str_list = xrealloc(
+                str_list, (strlen(str_list) + (li->tag ? strlen(li->tag) : 0) + 10)
+            );
+            (void) strcat(str_list, (li->tag ? li->tag : "HEAD"));
+            break;
 		/* All other characters, we insert an empty field (but
 		   we do put in the comma separating it from other
 		   fields).  This way if future CVS versions add formatting
