--- cvs-1.11.23/src/cvs.h	2018-02-19 17:00:07.063068131 +0000
+++ cvs-1.11.23/src/cvs.h.old	2018-02-19 17:00:27.875182393 +0000
@@ -245,7 +245,7 @@
 #define CVSEDITPREFIXLEN 4
 
 #define	CVSLCKAGE	(60*60)		/* 1-hour old lock files cleaned up */
-#define	CVSLCKSLEEP	30		/* wait 30 seconds before retrying */
+#define	CVSLCKSLEEP	3		/* wait 30 seconds before retrying */
 #define	CVSBRANCH	"1.1.1"		/* RCS branch used for vendor srcs */
 
 #ifdef USE_VMS_FILENAMES
