diff --git salt/modules/aptpkg.py salt/modules/aptpkg.py
index 7ded2944c..b29d3d972 100644
--- salt/modules/aptpkg.py
+++ salt/modules/aptpkg.py
@@ -165,16 +165,28 @@ if not HAS_APT:
             Parse lines from sources files
             """
             self.disabled = False
-            repo_line = self.line.strip().split()
-            if not repo_line:
+            line = self.line.strip()
+            if not line:
                 self.invalid = True
                 return False
-            if repo_line[0].startswith("#"):
-                repo_line.pop(0)
+
+            if line.startswith("#"):
                 self.disabled = True
-            if repo_line[0] not in ["deb", "deb-src", "rpm", "rpm-src"]:
+                repo_line = line[1:]
+
+            idx = line.find("#")
+            if idx > 0:
+                self.comment = line[idx + 1 :]
+                line = line[:idx]
+
+            repo_line = line.strip().split()
+            if (
+                repo_line[0] not in ["deb", "deb-src", "rpm", "rpm-src"]
+                or len(repo_line) < 3
+            ):
                 self.invalid = True
                 return False
+
             if repo_line[1].startswith("["):
                 opts = re.search(r"\[.*\]", self.line).group(0).strip("[]")
                 repo_line = [x for x in (line.strip("[]") for line in repo_line) if x]
