diff --git salt/minion.py salt/minion.py
index 63fb6e91ea..d5b4cbcdf9 100644
--- salt/minion.py
+++ salt/minion.py
@@ -445,14 +445,18 @@ class MinionBase:
         """
         if context is None:
             context = {}
+
         if initial_load:
-            self.opts["pillar"] = salt.pillar.get_pillar(
-                self.opts,
-                self.opts["grains"],
-                self.opts["id"],
-                self.opts["saltenv"],
-                pillarenv=self.opts.get("pillarenv"),
-            ).compile_pillar()
+            if self.opts.get('ignore_init_pillars'):
+                self.opts['pillar'] = {}
+            else:
+                self.opts["pillar"] = salt.pillar.get_pillar(
+                    self.opts,
+                    self.opts["grains"],
+                    self.opts["id"],
+                    self.opts["saltenv"],
+                    pillarenv=self.opts.get("pillarenv"),
+                ).compile_pillar()
 
         self.utils = salt.loader.utils(self.opts, context=context)
         self.functions = salt.loader.minion_mods(
