diff --git salt/utils/jinja.py salt/utils/jinja.py
index 0cb70bf64a..c0c1997a6d 100644
--- salt/utils/jinja.py
+++ salt/utils/jinja.py
@@ -3,7 +3,6 @@ Jinja loading utils to enable a more powerful backend for jinja templates
 """
 
 
-import atexit
 import logging
 import os.path
 import pipes
@@ -198,7 +197,7 @@ class SaltCacheLoader(BaseLoader):
         raise TemplateNotFound(template)
 
 
-atexit.register(SaltCacheLoader.shutdown)
+# atexit.register(SaltCacheLoader.shutdown)
 
 
 class PrintableDict(OrderedDict):
