diff --git salt/modules/lxd.py salt/modules/lxd.py
index 8f5ff6044c..c855234a89 100644
--- salt/modules/lxd.py
+++ salt/modules/lxd.py
@@ -3628,22 +3628,21 @@ if HAS_PYLXD:
     except ImportError:
         from pylxd.models.container import Container
 
-    class FilesManager(Container.FilesManager):
-        def put(self, filepath, data, mode=None, uid=None, gid=None):
-            headers = {}
-            if mode is not None:
-                if isinstance(mode, int):
-                    mode = oct(mode)
-                elif not mode.startswith("0"):
-                    mode = "0{}".format(mode)
-                headers["X-LXD-mode"] = mode
-            if uid is not None:
-                headers["X-LXD-uid"] = str(uid)
-            if gid is not None:
-                headers["X-LXD-gid"] = str(gid)
-            response = self._client.api.containers[self._container.name].files.post(
-                params={"path": filepath}, data=data, headers=headers
-            )
-            return response.status_code == 200
+    def put(self, filepath, data, mode=None, uid=None, gid=None):
+        headers = {}
+        if mode is not None:
+            if isinstance(mode, int):
+                mode = oct(mode)
+            elif not mode.startswith("0"):
+                mode = "0{}".format(mode)
+            headers["X-LXD-mode"] = mode
+        if uid is not None:
+            headers["X-LXD-uid"] = str(uid)
+        if gid is not None:
+            headers["X-LXD-gid"] = str(gid)
+        response = self._client.api.containers[self._container.name].files.post(
+            params={"path": filepath}, data=data, headers=headers
+        )
+        return response.status_code == 200
 
-    Container.FilesManager = FilesManager
+    Container.FilesManager.put = put
