diff --git salt/utils/yamlloader.py salt/utils/yamlloader.py
index a22ccd7fea7..9b5544d31ab 100644
--- salt/utils/yamlloader.py
+++ salt/utils/yamlloader.py
@@ -3,8 +3,6 @@ Custom YAML loading in Salt
 """
 
 
-import warnings
-
 import salt.utils.stringutils
 import yaml  # pylint: disable=blacklisted-import
 from yaml.constructor import ConstructorError
@@ -17,15 +15,6 @@ BaseLoader = getattr(yaml, "CSafeLoader", yaml.SafeLoader)
 __all__ = ["SaltYamlSafeLoader", "load", "safe_load"]
 
 
-class DuplicateKeyWarning(RuntimeWarning):
-    """
-    Warned when duplicate keys exist
-    """
-
-
-warnings.simplefilter("always", category=DuplicateKeyWarning)
-
-
 # with code integrated from https://gist.github.com/844388
 class SaltYamlSafeLoader(BaseLoader):
     """
diff --git salt/utils/yamlloader_old.py salt/utils/yamlloader_old.py
index a5b005e7dc4..68f2cfd3d54 100644
--- salt/utils/yamlloader_old.py
+++ salt/utils/yamlloader_old.py
@@ -4,7 +4,6 @@ Custom YAML loading in Salt
 
 
 import re
-import warnings
 
 import salt.utils.stringutils
 import yaml  # pylint: disable=blacklisted-import
@@ -21,15 +20,6 @@ except Exception:  # pylint: disable=broad-except
 __all__ = ["SaltYamlSafeLoader", "load", "safe_load"]
 
 
-class DuplicateKeyWarning(RuntimeWarning):
-    """
-    Warned when duplicate keys exist
-    """
-
-
-warnings.simplefilter("always", category=DuplicateKeyWarning)
-
-
 # with code integrated from https://gist.github.com/844388
 class SaltYamlSafeLoader(yaml.SafeLoader):
     """
