diff --git salt/states/kubernetes.py salt/states/kubernetes.py
index ed78f34..ec824cc 100644
--- salt/states/kubernetes.py
+++ salt/states/kubernetes.py
@@ -528,11 +528,13 @@ def secret_present(
             **kwargs
         )
 
-    ret["changes"] = {
-        # Omit values from the return. They are unencrypted
-        # and can contain sensitive data.
-        "data": list(res["data"])
-    }
+    diff = set(res["data"].items()) - set(secret.get("data", {}).items())
+    if diff:
+        ret["changes"] = {
+            # Omit values from the return. They are unencrypted
+            # and can contain sensitive data.
+            "data": set(res["data"]) - set(secret.get("data", {}))
+        }
     ret["result"] = True
 
     return ret
