diff --git a/yweb/robot/kiwi_queries/robot/triggers/portiontrigger/deploy.mq b/yweb/robot/kiwi_queries/robot/triggers/portiontrigger/deploy.mq
index 7e5b197..a74a871 100644
--- a/yweb/robot/kiwi_queries/robot/triggers/portiontrigger/deploy.mq
+++ b/yweb/robot/kiwi_queries/robot/triggers/portiontrigger/deploy.mq
@@ -78,6 +78,11 @@ Statements {
       Description: "TDataContainer"
     }
     Params {
+      Name: "sentHashResult"
+      Type: AT_BLOB
+      Description: "TDataContainer"
+    }
+    Params {
       Name: "indexingSourceResult"
       Type: AT_BLOB
       Description: "TDataContainer"
diff --git a/yweb/robot/kiwi_queries/robot/triggers/portiontrigger/portiontrigger.cpp b/yweb/robot/kiwi_queries/robot/triggers/portiontrigger/portiontrigger.cpp
index db98b99..d4231a4 100644
--- a/yweb/robot/kiwi_queries/robot/triggers/portiontrigger/portiontrigger.cpp
+++ b/yweb/robot/kiwi_queries/robot/triggers/portiontrigger/portiontrigger.cpp
@@ -128,6 +128,10 @@ public:
                 ythrow yexception() << "PortionTrigger: invalid type of the last input value, expected 'links' type AT_BLOB.";
         }

+        if (!extraDirectText) {
+            ythrow yexception() << "PortionTrigger: extraDirectText is missing (check SegmentatorTrigger!)";
+        }
+
         TBuffer arcBuf;
         {
             NIndexerCore::NIndexerCorePrivate::TDirectData data;
